docs/api/en/objects/BatchedMesh.html

```diff
@@ -222,22 +228,35 @@ <h3>
 			the length of the given geometry index buffer.
 		</p>
 		<p>
-			Adds the given geometry to the [name] and returns the associated index referring to it.
+			Adds the given geometry and a new instance to the [name] and returns the associated index referring to it to be used by other functions.
+		</p>
+
+		<h3>
+			[method:Integer addInstance]( [param:Integer id] )
+		</h3>
+		<p>
+			[page:Integer id]: The id of a previously added item via "addGeometry" or "addInstance" to add into the [name] to render. 
+		</p>
+		<p>
+			Adds a a new instance to the [name] using the same geometry from the given id and returns the associated index referring to it to be used
+			by other functions.
 		</p>
 
 		<h3>
 			[method:Integer setGeometryAt]( [param:Integer index], [param:BufferGeometry geometry] )
 		</h3>
 		<p>
-			[page:Integer index]: Which geometry index to replace with this geometry.
+			[page:Integer index]: Which item index to replace with this geometry.
 		</p>
 		<p>
 			[page:BufferGeometry geometry]: The geometry to substitute at the given geometry index.
 		</p>
 		<p>
-			Replaces the geometry at `index` with the provided geometry. Throws an error if there is not enough space reserved for geometry at the index.
+			Replaces the geometry used by the item at `index` with the provided geometry. Throws an error if there is not enough space reserved for geometry
+			at the index. Calling this will change all instances that are rendering that geometry.
```

Comment:
This line answers some of my questions above, but now I'm wondering. Is this by design, by technical requirement, or undecided?

Suppose a BatchedMesh represents a forest with instances of various tree species, each having a high-res and low-res geometry available. Trees within 100m use a high-poly geometry, and those further away use low-res geometries. When an instance crosses the 100m radius I might want to swap out the geometry of that instance alone, which may be difficult within this API. I expect it would be less common to want to the shared geometry of all instances at once, but I'm not sure.

Reply:
Yeah this was one of the reasons for the other notes in #28404:

> Another API option might be to enabling managing of geometry vs instances separately ...

There are technically two types of ids at being used here. A "geometry id", which identifies a range in the BatchedMesh geometry to draw, and an "instance id", which internally refers to a geometry to draw via geometry id. Instance ids are, of course, used to set matrices colors, etc of the given id, as well. The current API obscures geometry ids by referencing the geometries via instance id in the interest of making the API easier to understand but it may just make it more confusing in some cases.

In the LoD case you're referring to users could just hide and show associated with different LoDs but I agree that's not ideal if only because of memory usage (unnecessary matrix data storage). And adding / deleting instances will have a performance cost to "fix up" all ids & matrix textures once the "optimize" function is eventually added. We could add a "swap geometry" function to change the underlying geometry pointer but this will lead to other issues like losing the underlying geometry id once all the associated instances lose it.

I'm starting to feel that we should expose the underlying geometry ids (via `addGeometry`) in addition to instances (via `addInstance`) so the user can be aware of and control both, which I think may be more explainable, too. And then other class abstractions can simplify things further if needed. The downsides are that this will be a bit more verbose and be a breaking change. Using the class would now look like this (copied from the other PR):

```js
const mesh = new BatchedMesh( ... );

// initialize geometries
const geometryId0 = mesh.addGeometry( sphereGeometry );
const geometryId1 = mesh.addGeometry( boxGeometry );

// initialize instances
const instancedId0 = mesh.addInstance( geometryId0 );
mesh.setMatrixAt( instanceId0, ... );

const instancedId1 = mesh.addInstance( geometryId1 );
mesh.setMatrixAt( instanceId1, ... );

const instancedId2 = mesh.addInstance( geometryId0 );
mesh.setMatrixAt( instanceId2, ... );

const instancedId3 = mesh.addInstance( geometryId1 );
mesh.setMatrixAt( instanceId3, ... );
```

Thanks for thinking about some other use cases!

Reply:
> In the LoD case you're referring to users could just hide and show associated with different LoDs but I agree that's not ideal if only because of memory usage ...

Yes, it's currently a similar API InstancedMesh. To support a similar use case there we need multiple InstancedMesh objects, and move the instance matrix between them to 'change' that instance.  I think it can be confusing that users must move the instance matrix around to change the geometry of the instance, and this feels similar.

I like the API direction above. Maybe in the interest of getting a first version out sooner, what do you think of saying...

1. `addGeometry` returns a geometry ID
2. `addInstance` takes a geometry ID and returns an instance ID
3. instance IDs and geometry IDs are not interchangeable (at least for now)

.... and perhaps we do a separate PR for `setGeometryForInstance` to keep things simpler. I think the API `setGeometryAt( id, geometry )` could also work, we'd just need to keep a `WeakMap<BufferGeometry, number>` for the lookups, and there's probably a little overhead in that.

> The downsides are that this will be a bit more verbose and be a breaking change. 

What is the breaking change compared to today? Just that calling `addGeometry` alone doesn't cause anything to render?

Reply:
We could also keep the API as you first proposed, say that all IDs are instance IDs, and  change the behavior of `setGeometryAt` to affect only one instance. Either:

```javascript
// (A) geometry must already be in BatchedMesh, with or without instances
batch.setGeometryAt( id, geometry )

// (B) geometry must already be in BatchedMesh, with >=1 instance 
batch.setGeometryAt( targetId, sourceId );
```

If the last instance of a geometry is removed, all references are lost, and it's eligible for some hypothetical future cleanup process.

Reply:
> - addGeometry returns a geometry ID
> - addInstance takes a geometry ID and returns an instance ID
> - instance IDs and geometry IDs are not interchangeable (at least for now)

Yes this is what I would propose for now if we're going to change the API. We can add more functions later.

> I think the API setGeometryAt( id, geometry ) could also work

Perhaps as an option? But it shouldn't be required since once the geometry is in the batched mesh you should want to be able to discard / GC the original geometry to save memory. And this can also be done by a user abstraction.

> What is the breaking change compared to today? That calling addGeometry alone doesn't cause anything to render?

Yeah - just this. The user must now call "addInstance" after "addGeometry".

> If the last instance of a geometry is removed, all references are lost, and it's eligible for some hypothetical future cleanup process.

This is basically what's happening now. But I don't think I love the idea of requiring that users keep geometry around once it's been added to the BatchedMesh. As I mention above I think one expected benefit is that you can GC it after.

I'm inclined to change the API to require calling `addInstance` if that sounds good to you?

Reply:
I've updated the docs to reflect the separation of instanceId and geometryId.

Reply:
Agreed on all points! 👍